www_src/pages/project/Positionable.jsx

```diff
@@ -0,0 +1,135 @@
+var Positionable = React.createClass({
+
+  getInitialState: function() {
+    return {
+      x: this.props.x || 0,
+      y: this.props.y || 0,
+      scale: this.props.scale || 1,
+      angle: this.props.angle || 0,
+      zIndex: (typeof this.props.zIndex !== "undefined") ? this.props.zIndex : 1,
+      interactive: (typeof this.props.interactive !== "undefined") ? this.props.interactive : true,
+      touchactive: false
+    };
+  },
+
+  componenWillMount: function() {
+    if(!this.state.interactive) return;
+
+    this.markx = this.state.x;
+    this.marky = this.state.y;
+    this.markangle = this.state.angle;
+    this.markscale = this.state.scale;
+  },
+
+  componentDidMount: function() {
+    if(!this.state.interactive) return;
+    var touchHandler = this.touchhandler = require("./touchhandler")(this);
+
+    var dnode = this.getDOMNode();
+    dnode.addEventListener("touchstart", touchHandler.startmark);
+    dnode.addEventListener("mousedown", touchHandler.startmark);
+    dnode.addEventListener("touchmove", touchHandler.panmove);
+    dnode.addEventListener("mousemove", touchHandler.panmove);
+    dnode.addEventListener("touchend", touchHandler.endmark);
+    dnode.addEventListener("mouseup", touchHandler.endmark);
+
+    var onode = this.refs.overlay.getDOMNode();
+    onode.addEventListener("touchstart", touchHandler.secondFinger);
+    onode.addEventListener("touchmove", touchHandler.handleRS);
+    onode.addEventListener("touchend", touchHandler.endSecondFinger);
+  },
+
+  componentDidUpdate: function(prevProps, prevState) {
+    if(!prevState.touchactive && this.state.touchactive) {
+      this.props.onUpdate(this.state);
+    }
+  },
+
+  render: function() {
+    var x = this.state.x,
+        y = this.state.y,
+        angle = this.state.angle,
+        scale = this.state.scale,
+        zIndex = this.state.zIndex;
+
+    var style = {
+      transform: [
+        "translate("+x+"px, "+y+"px)",
+        "rotate("+angle+"deg)",
+        "scale("+scale+")"
+      ].join(" "),
+      transformOrigin: "center",
+      zIndex: zIndex
+    };
+
+    var className = classes({
+      positionable: true,
+      touchactive: this.state.touchactive,
+      current: this.props.current
+    });
+
+    return (
+      <div>
+        <div ref="overlay" className="touchOverlay" hidden={!this.state.touchactive} />
+        <div style={style} className={className}>
+          { this.props.children }
+        </div>
+      </div>
+    );
+  },
+
+  handleTranslation: function(x, y) {
+    this.setState({
+      x: this.transform.x + x,
+      y: this.transform.y + y
+    }, function() {
+      if(this.props.onUpdate) {
+        this.props.onUpdate(this.state);
+      }
+    });
+  },
+
+  handleRotationAndScale: function(angle, scale) {
+    this.setState({
+      angle: this.transform.angle + angle,
+      scale: this.transform.scale * scale
+    }, function() {
+      if(this.props.onUpdate) {
+        this.props.onUpdate(this.state);
```

Comment:
turns out I had a reason for this and forgot - a general purpose update in componentDidUpdate will cause infinite recursions, because this.props.onUpdate causes a rerender in Project, which triggers a componentDidUpdate because of changed props. In this case, the update should only kick in for changed state, so they're triggered specifically as a result of setState